src/diffusers/pipelines/marigold/pipeline_marigold_normals.py

```diff
@@ -0,0 +1,648 @@
+        latent (`None`, `torch.FloatTensor`):
+            Latent features corresponding to the predictions, compatible with the `latents` argument of the pipeline.
+            The shape is $numimages * numensemble \times 4 \times latentheight \times latentwidth$.
+    """
+
+    prediction: Union[np.ndarray, torch.FloatTensor]
+    uncertainty: Union[None, np.ndarray, torch.FloatTensor]
+    latent: Union[None, torch.FloatTensor]
+
+
+class MarigoldNormalsPipeline(DiffusionPipeline):
+    """
+    Pipeline for monocular normals estimation using the Marigold method: https://marigoldmonodepth.github.io.
+
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+
+    Args:
+        unet (`UNet2DConditionModel`):
+            Conditional U-Net to denoise the normals latent, conditioned on image latent.
+        vae (`AutoencoderKL`):
+            Variational Auto-Encoder (VAE) Model to encode and decode images and predictions to and from latent
+            representations.
+        scheduler (`DDIMScheduler` or `LCMScheduler`):
+            A scheduler to be used in combination with `unet` to denoise the encoded image latents.
+        text_encoder (`CLIPTextModel`):
+            Text-encoder, for empty text embedding.
+        tokenizer (`CLIPTokenizer`):
+            CLIP tokenizer.
+        prediction_type (`str`, *optional*):
+            Type of predictions made by the model.
+        use_full_z_range (`bool`, *optional*):
+            Whether the normals predicted by this model utilize the full range of the Z dimension, or only its positive
+            half.
+        default_denoising_steps (`int`, *optional*):
+            The minimum number of denoising diffusion steps that are required to produce a prediction of reasonable
+            quality with the given model. This value must be set in the model config. When the pipeline is called
+            without explicitly setting `num_inference_steps`, the default value is used. This is required to ensure
+            reasonable results with various model flavors compatible with the pipeline, such as those relying on very
+            short denoising schedules (`LCMScheduler`) and those with full diffusion schedules (`DDIMScheduler`).
+        default_processing_resolution (`int`, *optional*):
+            The recommended value of the `processing_resolution` parameter of the pipeline. This value must be set in
+            the model config. When the pipeline is called without explicitly setting `processing_resolution`, the
+            default value is used. This is required to ensure reasonable results with various model flavors trained
+            with varying optimal processing resolution values.
+    """
+
+    model_cpu_offload_seq = "text_encoder->vae.encoder->unet->vae.decoder"
```

Comment:
Here it is actually quite important that the sequential offloading happens in this order: text_encoder->vae.encoder->unet->vae.decoder. Can someone confirm if this is the case? Both pipelines invoke modules in exactly this order

Reply:
I don't think it would work - the hooks should be installed on the model, not part of the model https://github.com/huggingface/diffusers/blob/0f0defdb652d24cd979317e70d17ca213353bfde/src/diffusers/pipelines/pipeline_utils.py#L1072

Reply:
if you do `model_cpu_offload_seq = "text_encoder->unet->vae"`
it will be moved to GPU first time it's called for encoding and stay there until the end

Reply:
If `vae` is in the chain, the encoding phase fails in the first convolution of the vae encoder with the error that the module is on CPU and the input is on GPU.

Reply:
`text_encoder -> unet -> vae` would not work? 
that's what we do for other pipelines with similar situation e.g. https://github.com/huggingface/diffusers/blob/6529ee67ec02fcf58d2fd9242164ea002b351d75/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L214

Reply:
ohh I see

Reply:
Can you refactor your `prepare_latents` method to be same as ours with minimum change? https://github.com/huggingface/diffusers/blob/6529ee67ec02fcf58d2fd9242164ea002b351d75/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L727
most importantly, we need to use `vae.encode` method (https://github.com/huggingface/diffusers/blob/6529ee67ec02fcf58d2fd9242164ea002b351d75/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L749) to be able to offload vae encoder correctly ( to leverage the `apply_forward_hook` decorator https://github.com/huggingface/diffusers/blob/6529ee67ec02fcf58d2fd9242164ea002b351d75/src/diffusers/utils/accelerate_utils.py#L27)

Reply:
with this change, you will be able to use this sequence `text_encoder -> unet -> vae`
vae will be moved to cuda first time it is called (after we're done with text_encoder) and it will stay there until the end, after `decode_latent`

Reply:
How does https://github.com/huggingface/diffusers/blob/6529ee67ec02fcf58d2fd9242164ea002b351d75/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L77-L87 go with `torch.compile`-friendliness? We had to use `return_dict=False` here `prediction = self.vae.decode(pred_latent / self.vae.config.scaling_factor, return_dict=False)[0]  # [B,3,H,W]` in `decode_prediction` to promote that

Reply:
Done otherwise, thanks for the guidance!